File: packages/kotlin/src/main/kotlin/dev/tempera/sdk/Json.kt
// A minimal, order-preserving JSON model.
//
// The package has no external dependencies -- the JDK ships no JSON API -- and
// the wire contract needs more than a generic one anyway: request bodies must
// serialize their members in the order the surface tables declare them (so one
// request is one reproducible byte string, retried verbatim), and error bodies
// must be read without losing member order or number precision. This file is
// the Kotlin counterpart of the private scanner in
// `packages/rust/src/error.rs`.

package dev.tempera.sdk

/** One member of a JSON object, in declaration order. */
public data class TemperaJsonMember(
    public val key: String,
    public val value: TemperaJson,
)

/**
 * A JSON value. Objects keep their members in order, so a body built from the
 * surface tables serializes the same way every time.
 */
public sealed class TemperaJson {
    /** JSON `null`. */
    public object Null : TemperaJson()

    /** JSON `true` / `false`. */
    public data class Bool(public val value: Boolean) : TemperaJson()

    /** A JSON number with no fractional or exponent part. */
    public data class Int64(public val value: Long) : TemperaJson()

    /** A JSON number with a fractional or exponent part. */
    public data class Decimal(public val value: Double) : TemperaJson()

    /** A JSON string. */
    public data class Text(public val value: String) : TemperaJson()

    /** A JSON array. */
    public data class Arr(public val values: List<TemperaJson>) : TemperaJson()

    /** A JSON object, in member order. */
    public data class Obj(public val members: List<TemperaJsonMember>) : TemperaJson()

    /** The value of one object member, or `null` for a non-object or a missing key. */
    public operator fun get(key: String): TemperaJson? =
        if (this is Obj) members.firstOrNull { it.key == key }?.value else null

    /** The string payload, when this is a string. */
    public fun asString(): String? = if (this is Text) value else null

    /** The integer payload, when this is a number that is exactly an integer. */
    public fun asLong(): Long? =
        when (this) {
            is Int64 -> value
            is Decimal ->
                if (value.isFinite() && value == Math.floor(value)) value.toLong() else null
            else -> null
        }

    /** The integer payload as an `Int`. */
    public fun asInt(): Int? = asLong()?.toInt()

    /** The boolean payload, when this is a boolean. */
    public fun asBoolean(): Boolean? = if (this is Bool) value else null

    /** The elements, when this is an array. */
    public fun asArray(): List<TemperaJson>? = if (this is Arr) values else null

    /** The members, when this is an object. */
    public fun asObject(): List<TemperaJsonMember>? = if (this is Obj) members else null

    /** Whether this is JSON `null`. */
    public fun isNull(): Boolean = this === Null

    /**
     * The plain-text form used for path substitution and query-string values.
     *
     * Strings are used as-is; numbers and booleans take their JSON spelling,
     * which is what the TypeScript and Python clients send.
     */
    public fun plainText(): String =
        when (this) {
            is Null -> ""
            is Bool -> if (value) "true" else "false"
            is Int64 -> value.toString()
            is Decimal -> numberText(value)
            is Text -> value
            is Arr, is Obj -> serialized()
        }

    /** Serialize compactly, with no insignificant whitespace, in member order. */
    public fun serialized(): String {
        val out = StringBuilder()
        writeTo(out)
        return out.toString()
    }

    /** UTF-8 bytes of the compact serialization. */
    public fun serializedBytes(): ByteArray = serialized().toByteArray(Charsets.UTF_8)

    private fun writeTo(out: StringBuilder) {
        when (this) {
            is Null -> out.append("null")
            is Bool -> out.append(if (value) "true" else "false")
            is Int64 -> out.append(value.toString())
            is Decimal -> out.append(numberText(value))
            is Text -> {
                out.append('"')
                escapeInto(value, out)
                out.append('"')
            }
            is Arr -> {
                out.append('[')
                for ((index, element) in values.withIndex()) {
                    if (index > 0) out.append(',')
                    element.writeTo(out)
                }
                out.append(']')
            }
            is Obj -> {
                out.append('{')
                for ((index, member) in members.withIndex()) {
                    if (index > 0) out.append(',')
                    out.append('"')
                    escapeInto(member.key, out)
                    out.append("\":")
                    member.value.writeTo(out)
                }
                out.append('}')
            }
        }
    }

    public companion object {
        /**
         * Escape one string for a JSON string literal, without the surrounding
         * quotes. Non-ASCII characters are emitted verbatim as UTF-8, matching
         * the compact `ensure_ascii=False` encoder in the Python package.
         */
        public fun escape(value: String): String {
            val out = StringBuilder(value.length)
            escapeInto(value, out)
            return out.toString()
        }

        /**
         * Parse a complete JSON document, or `null` on any syntax error or
         * trailing garbage -- which callers treat as "unparseable body".
         */
        public fun parse(input: String): TemperaJson? {
            val scanner = TemperaJsonScanner(input)
            scanner.skipWhitespace()
            val value = scanner.parseValue() ?: return null
            scanner.skipWhitespace()
            return if (scanner.isAtEnd()) value else null
        }

        /** Parse UTF-8 bytes; `null` when they are not valid JSON. */
        public fun parse(input: ByteArray): TemperaJson? =
            if (input.isEmpty()) null else parse(String(input, Charsets.UTF_8))
    }
}

internal fun escapeInto(value: String, out: StringBuilder) {
    for (character in value) {
        when (character) {
            '"' -> out.append("\\\"")
            '\\' -> out.append("\\\\")
            '\n' -> out.append("\\n")
            '\r' -> out.append("\\r")
            '\t' -> out.append("\\t")
            else ->
                if (character.code < 0x20) {
                    out.append(String.format("\\u%04x", character.code))
                } else {
                    out.append(character)
                }
        }
    }
}

internal fun numberText(value: Double): String {
    if (!value.isFinite()) return "null"
    if (value == Math.floor(value) && Math.abs(value) < 1e15) {
        return value.toLong().toString()
    }
    return value.toString()
}

/**
 * Convert a Kotlin value into JSON: `null`, `Boolean`, any integral or
 * floating-point `Number`, `CharSequence`, `Map`, `Iterable`, `Array`, or an
 * already-built [TemperaJson]. Anything else takes its `toString()`.
 */
public fun temperaJsonOf(value: Any?): TemperaJson =
    when (value) {
        null -> TemperaJson.Null
        is TemperaJson -> value
        is Boolean -> TemperaJson.Bool(value)
        is Byte, is Short, is Int, is Long -> TemperaJson.Int64((value as Number).toLong())
        is Float, is Double -> TemperaJson.Decimal((value as Number).toDouble())
        is CharSequence -> TemperaJson.Text(value.toString())
        is Map<*, *> ->
            TemperaJson.Obj(
                value.entries.map { TemperaJsonMember(it.key.toString(), temperaJsonOf(it.value)) }
            )
        is Iterable<*> -> TemperaJson.Arr(value.map { temperaJsonOf(it) })
        is Array<*> -> TemperaJson.Arr(value.map { temperaJsonOf(it) })
        else -> TemperaJson.Text(value.toString())
    }

/** Build a JSON object from ordered pairs. */
public fun temperaJsonObject(vararg members: Pair<String, Any?>): TemperaJson =
    TemperaJson.Obj(members.map { TemperaJsonMember(it.first, temperaJsonOf(it.second)) })

/** A recursive-descent JSON scanner. */
internal class TemperaJsonScanner(private val source: String) {
    private var position = 0

    fun isAtEnd(): Boolean = position >= source.length

    private fun peek(): Char? = if (position < source.length) source[position] else null

    private fun bump(): Char? {
        val character = peek() ?: return null
        position += 1
        return character
    }

    fun skipWhitespace() {
        while (true) {
            val character = peek() ?: return
            if (character == ' ' || character == '\t' || character == '\n' || character == '\r') {
                position += 1
            } else {
                return
            }
        }
    }

    private fun eat(token: String): Boolean {
        if (source.startsWith(token, position)) {
            position += token.length
            return true
        }
        return false
    }

    fun parseValue(): TemperaJson? {
        val character = peek() ?: return null
        return when {
            character == '{' -> parseObject()
            character == '[' -> parseArray()
            character == '"' -> parseString()?.let { TemperaJson.Text(it) }
            character == 't' -> if (eat("true")) TemperaJson.Bool(true) else null
            character == 'f' -> if (eat("false")) TemperaJson.Bool(false) else null
            character == 'n' -> if (eat("null")) TemperaJson.Null else null
            character == '-' || (character in '0'..'9') -> parseNumber()
            else -> null
        }
    }

    private fun parseObject(): TemperaJson? {
        bump() // consume '{'
        val members = ArrayList<TemperaJsonMember>()
        skipWhitespace()
        if (peek() == '}') {
            bump()
            return TemperaJson.Obj(members)
        }
        while (true) {
            skipWhitespace()
            if (peek() != '"') return null
            val key = parseString() ?: return null
            skipWhitespace()
            if (bump() != ':') return null
            skipWhitespace()
            val value = parseValue() ?: return null
            members.add(TemperaJsonMember(key, value))
            skipWhitespace()
            when (bump()) {
                ',' -> continue
                '}' -> return TemperaJson.Obj(members)
                else -> return null
            }
        }
    }

    private fun parseArray(): TemperaJson? {
        bump() // consume '['
        val values = ArrayList<TemperaJson>()
        skipWhitespace()
        if (peek() == ']') {
            bump()
            return TemperaJson.Arr(values)
        }
        while (true) {
            skipWhitespace()
            val value = parseValue() ?: return null
            values.add(value)
            skipWhitespace()
            when (bump()) {
                ',' -> continue
                ']' -> return TemperaJson.Arr(values)
                else -> return null
            }
        }
    }

    private fun parseString(): String? {
        bump() // consume the opening quote
        val out = StringBuilder()
        while (true) {
            val character = bump() ?: return null
            if (character == '"') return out.toString()
            if (character.code < 0x20) return null // raw control character
            if (character != '\\') {
                out.append(character)
                continue
            }
            when (bump()) {
                '"' -> out.append('"')
                '\\' -> out.append('\\')
                '/' -> out.append('/')
                'b' -> out.append('\b')
                'f' -> out.append('\u000C')
                'n' -> out.append('\n')
                'r' -> out.append('\r')
                't' -> out.append('\t')
                'u' -> {
                    val unit = parseHex4() ?: return null
                    if (unit in 0xD800..0xDBFF) {
                        // High surrogate: a \uXXXX low surrogate must follow.
                        if (bump() != '\\' || bump() != 'u') return null
                        val low = parseHex4() ?: return null
                        if (low !in 0xDC00..0xDFFF) return null
                        out.append(Char(unit))
                        out.append(Char(low))
                    } else if (unit in 0xDC00..0xDFFF) {
                        return null // a lone low surrogate
                    } else {
                        out.append(Char(unit))
                    }
                }
                else -> return null
            }
        }
    }

    private fun parseHex4(): Int? {
        var value = 0
        for (index in 0 until 4) {
            val character = bump() ?: return null
            val digit = Character.digit(character, 16)
            if (digit < 0) return null
            value = (value shl 4) or digit
        }
        return value
    }

    private fun parseNumber(): TemperaJson? {
        val start = position
        var isInteger = true
        if (peek() == '-') position += 1
        if (!eatDigits()) return null
        if (peek() == '.') {
            isInteger = false
            position += 1
            if (!eatDigits()) return null
        }
        val exponent = peek()
        if (exponent == 'e' || exponent == 'E') {
            isInteger = false
            position += 1
            val sign = peek()
            if (sign == '+' || sign == '-') position += 1
            if (!eatDigits()) return null
        }
        val raw = source.substring(start, position)
        if (isInteger) {
            val parsed = raw.toLongOrNull()
            if (parsed != null) return TemperaJson.Int64(parsed)
        }
        val parsed = raw.toDoubleOrNull() ?: return null
        return TemperaJson.Decimal(parsed)
    }

    private fun eatDigits(): Boolean {
        val start = position
        while (true) {
            val character = peek() ?: break
            if (character in '0'..'9') position += 1 else break
        }
        return position > start
    }
}

File: packages/swift/Sources/TemperaSDK/Json.swift
// A minimal, order-preserving JSON model.
//
// The package has no external dependencies, and `JSONSerialization` cannot
// express what the wire contract needs: request bodies must serialize their
// members in the order the surface tables declare them (so one request is one
// reproducible byte string, retried verbatim), and error bodies must be read
// without losing member order or number precision. This file is the Swift
// counterpart of the private scanner in `packages/rust/src/error.rs`.

import Foundation

/// One member of a JSON object, in declaration order.
public struct TemperaJSONMember: Sendable, Equatable {
    /// The member's key.
    public let key: String
    /// The member's value.
    public let value: TemperaJSON

    /// Create one member.
    public init(_ key: String, _ value: TemperaJSON) {
        self.key = key
        self.value = value
    }
}

/// A JSON value. Objects keep their members in order, so a body built from the
/// surface tables serializes the same way every time.
public enum TemperaJSON: Sendable, Equatable {
    /// JSON `null`.
    case null
    /// JSON `true` / `false`.
    case bool(Bool)
    /// A JSON number with no fractional or exponent part.
    case int(Int)
    /// A JSON number with a fractional or exponent part.
    case double(Double)
    /// A JSON string.
    case string(String)
    /// A JSON array.
    case array([TemperaJSON])
    /// A JSON object, in member order.
    case object([TemperaJSONMember])
}

// TemperaJSON deliberately does NOT conform to ExpressibleByNilLiteral: it
// would make a bare `nil` in any TemperaJSON context mean `.null`, which
// silently turns `optional ?? nil` and `condition ? value : nil` into a
// non-optional `.null`. Write `.null` where JSON null is meant.

extension TemperaJSON: ExpressibleByBooleanLiteral {
    public init(booleanLiteral value: Bool) { self = .bool(value) }
}

extension TemperaJSON: ExpressibleByIntegerLiteral {
    public init(integerLiteral value: Int) { self = .int(value) }
}

extension TemperaJSON: ExpressibleByFloatLiteral {
    public init(floatLiteral value: Double) { self = .double(value) }
}

extension TemperaJSON: ExpressibleByStringLiteral {
    public init(stringLiteral value: String) { self = .string(value) }
}

extension TemperaJSON: ExpressibleByArrayLiteral {
    public init(arrayLiteral elements: TemperaJSON...) { self = .array(elements) }
}

extension TemperaJSON: ExpressibleByDictionaryLiteral {
    public init(dictionaryLiteral elements: (String, TemperaJSON)...) {
        self = .object(elements.map { TemperaJSONMember($0.0, $0.1) })
    }
}

extension TemperaJSON {
    /// The value of one object member, or `nil` for a non-object or a missing key.
    public subscript(key: String) -> TemperaJSON? {
        guard case let .object(members) = self else { return nil }
        return members.first { $0.key == key }?.value
    }

    /// The string payload, when this is a string.
    public var stringValue: String? {
        guard case let .string(value) = self else { return nil }
        return value
    }

    /// The integer payload, when this is a number that is exactly an integer.
    public var intValue: Int? {
        switch self {
        case let .int(value): return value
        case let .double(value) where value.rounded() == value && value.isFinite: return Int(value)
        default: return nil
        }
    }

    /// The boolean payload, when this is a boolean.
    public var boolValue: Bool? {
        guard case let .bool(value) = self else { return nil }
        return value
    }

    /// The elements, when this is an array.
    public var arrayValue: [TemperaJSON]? {
        guard case let .array(values) = self else { return nil }
        return values
    }

    /// The members, when this is an object.
    public var objectValue: [TemperaJSONMember]? {
        guard case let .object(members) = self else { return nil }
        return members
    }

    /// Whether this is JSON `null`.
    public var isNull: Bool {
        if case .null = self { return true }
        return false
    }

    /// The plain-text form used for path substitution and query-string values.
    ///
    /// Strings are used as-is; numbers and booleans take their JSON spelling,
    /// which is what the TypeScript and Python clients send.
    public var plainText: String {
        switch self {
        case .null: return ""
        case let .bool(value): return value ? "true" : "false"
        case let .int(value): return String(value)
        case let .double(value): return TemperaJSON.numberText(value)
        case let .string(value): return value
        case .array, .object: return serialized()
        }
    }

    static func numberText(_ value: Double) -> String {
        guard value.isFinite else { return "null" }
        if value.rounded() == value, abs(value) < 1e15 {
            return String(Int64(value))
        }
        return String(value)
    }

    /// Escape one string for inclusion in a JSON string literal, without the
    /// surrounding quotes. Non-ASCII characters are emitted verbatim as UTF-8,
    /// matching the compact `ensure_ascii=False` encoder in the Python package.
    public static func escape(_ value: String) -> String {
        var out = String()
        out.reserveCapacity(value.count)
        for character in value.unicodeScalars {
            switch character {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            default:
                if character.value < 0x20 {
                    out += String(format: "\\u%04x", character.value)
                } else {
                    out.unicodeScalars.append(character)
                }
            }
        }
        return out
    }

    /// Serialize compactly, with no insignificant whitespace and members in
    /// declaration order.
    public func serialized() -> String {
        switch self {
        case .null:
            return "null"
        case let .bool(value):
            return value ? "true" : "false"
        case let .int(value):
            return String(value)
        case let .double(value):
            return TemperaJSON.numberText(value)
        case let .string(value):
            return "\"\(TemperaJSON.escape(value))\""
        case let .array(values):
            return "[" + values.map { $0.serialized() }.joined(separator: ",") + "]"
        case let .object(members):
            let inner = members
                .map { "\"\(TemperaJSON.escape($0.key))\":\($0.value.serialized())" }
                .joined(separator: ",")
            return "{\(inner)}"
        }
    }

    /// UTF-8 bytes of the compact serialization.
    public func serializedData() -> Data {
        Data(serialized().utf8)
    }

    /// Parse a complete JSON document, or `nil` on any syntax error or trailing
    /// garbage -- which callers treat as "unparseable body".
    public static func parse(_ input: String) -> TemperaJSON? {
        var scanner = TemperaJSONScanner(Array(input.unicodeScalars))
        scanner.skipWhitespace()
        guard let value = scanner.parseValue() else { return nil }
        scanner.skipWhitespace()
        return scanner.isAtEnd ? value : nil
    }

    /// Parse UTF-8 bytes; `nil` when they are not valid UTF-8 or not valid JSON.
    public static func parse(_ data: Data) -> TemperaJSON? {
        guard let text = String(data: data, encoding: .utf8) else { return nil }
        return parse(text)
    }
}

/// A recursive-descent JSON scanner over Unicode scalars.
struct TemperaJSONScanner {
    private let scalars: [Unicode.Scalar]
    private var position = 0

    init(_ scalars: [Unicode.Scalar]) {
        self.scalars = scalars
    }

    var isAtEnd: Bool { position >= scalars.count }

    private func peek() -> Unicode.Scalar? {
        position < scalars.count ? scalars[position] : nil
    }

    private mutating func bump() -> Unicode.Scalar? {
        guard let scalar = peek() else { return nil }
        position += 1
        return scalar
    }

    mutating func skipWhitespace() {
        while let scalar = peek(), scalar == " " || scalar == "\t" || scalar == "\n" || scalar == "\r" {
            position += 1
        }
    }

    private mutating func eat(_ token: String) -> Bool {
        let expected = Array(token.unicodeScalars)
        guard position + expected.count <= scalars.count else { return false }
        for (offset, scalar) in expected.enumerated() where scalars[position + offset] != scalar {
            return false
        }
        position += expected.count
        return true
    }

    mutating func parseValue() -> TemperaJSON? {
        guard let scalar = peek() else { return nil }
        switch scalar {
        case "{": return parseObject()
        case "[": return parseArray()
        case "\"": return parseString().map { TemperaJSON.string($0) }
        case "t": return eat("true") ? .bool(true) : nil
        case "f": return eat("false") ? .bool(false) : nil
        case "n": return eat("null") ? .null : nil
        default:
            if scalar == "-" || (scalar.value >= 0x30 && scalar.value <= 0x39) { return parseNumber() }
            return nil
        }
    }

    private mutating func parseObject() -> TemperaJSON? {
        _ = bump()  // consume '{'
        var members: [TemperaJSONMember] = []
        skipWhitespace()
        if peek() == "}" {
            _ = bump()
            return .object(members)
        }
        while true {
            skipWhitespace()
            guard peek() == "\"", let key = parseString() else { return nil }
            skipWhitespace()
            guard bump() == ":" else { return nil }
            skipWhitespace()
            guard let value = parseValue() else { return nil }
            members.append(TemperaJSONMember(key, value))
            skipWhitespace()
            switch bump() {
            case ",": continue
            case "}": return .object(members)
            default: return nil
            }
        }
    }

    private mutating func parseArray() -> TemperaJSON? {
        _ = bump()  // consume '['
        var values: [TemperaJSON] = []
        skipWhitespace()
        if peek() == "]" {
            _ = bump()
            return .array(values)
        }
        while true {
            skipWhitespace()
            guard let value = parseValue() else { return nil }
            values.append(value)
            skipWhitespace()
            switch bump() {
            case ",": continue
            case "]": return .array(values)
            default: return nil
            }
        }
    }

    private mutating func parseString() -> String? {
        _ = bump()  // consume the opening quote
        var out = String.UnicodeScalarView()
        while true {
            guard let scalar = bump() else { return nil }
            if scalar == "\"" { return String(out) }
            if scalar.value < 0x20 { return nil }  // raw control character
            guard scalar == "\\" else {
                out.append(scalar)
                continue
            }
            guard let escape = bump() else { return nil }
            switch escape {
            case "\"": out.append("\"")
            case "\\": out.append("\\")
            case "/": out.append("/")
            case "b": out.append(Unicode.Scalar(UInt8(0x08)))
            case "f": out.append(Unicode.Scalar(UInt8(0x0C)))
            case "n": out.append("\n")
            case "r": out.append("\r")
            case "t": out.append("\t")
            case "u":
                guard let unit = parseHex4() else { return nil }
                if unit >= 0xD800, unit < 0xDC00 {
                    // High surrogate: a \uXXXX low surrogate must follow.
                    guard bump() == "\\", bump() == "u", let low = parseHex4(),
                        low >= 0xDC00, low < 0xE000
                    else { return nil }
                    let combined =
                        0x10000 + ((UInt32(unit) - 0xD800) << 10) + (UInt32(low) - 0xDC00)
                    guard let scalar = Unicode.Scalar(combined) else { return nil }
                    out.append(scalar)
                } else {
                    guard let scalar = Unicode.Scalar(UInt32(unit)) else { return nil }
                    out.append(scalar)
                }
            default:
                return nil
            }
        }
    }

    private mutating func parseHex4() -> UInt16? {
        var value: UInt16 = 0
        for _ in 0..<4 {
            guard let scalar = bump(), let digit = hexDigit(scalar) else { return nil }
            value = (value << 4) | UInt16(digit)
        }
        return value
    }

    private func hexDigit(_ scalar: Unicode.Scalar) -> UInt8? {
        switch scalar {
        case "0"..."9": return UInt8(scalar.value - 0x30)
        case "a"..."f": return UInt8(scalar.value - 0x61 + 10)
        case "A"..."F": return UInt8(scalar.value - 0x41 + 10)
        default: return nil
        }
    }

    private mutating func parseNumber() -> TemperaJSON? {
        let start = position
        var isInteger = true
        if peek() == "-" { position += 1 }
        guard eatDigits() else { return nil }
        if peek() == "." {
            isInteger = false
            position += 1
            guard eatDigits() else { return nil }
        }
        if let scalar = peek(), scalar == "e" || scalar == "E" {
            isInteger = false
            position += 1
            if let sign = peek(), sign == "+" || sign == "-" { position += 1 }
            guard eatDigits() else { return nil }
        }
        let raw = String(String.UnicodeScalarView(scalars[start..<position]))
        if isInteger, let value = Int(raw) { return .int(value) }
        guard let value = Double(raw) else { return nil }
        return .double(value)
    }

    private mutating func eatDigits() -> Bool {
        let start = position
        while let scalar = peek(), scalar.value >= 0x30, scalar.value <= 0x39 {
            position += 1
        }
        return position > start
    }
}
